### PR TITLE
MM-14723: Add config flag for creating bots

### LINKED ISSUE #10776: Add system console setting to enable/disable the creation of bot accounts using the REST API.
If you're interested please comment here and come [join our "Contributors" community channel](https://community.mattermost.com/core/channels/tickets) on our daily build server, where you can discuss questions with community members and the Mattermost core team. For technical advice or questions, please  [join our "Developers" community channel](https://community.mattermost.com/core/channels/developers).

New contributors please see our [Developer's Guide](https://developers.mattermost.com/contribute/getting-started/).

----

**Notes**: [Jira ticket](https://mattermost.atlassian.net/browse/MM-14723)

The setting will need to be added to the system console.

The bot accounts backstage UI should be hidden when creation is disabled.

*Title*: Enable Bot Accounts
*Options*: true/false (default: false)
*Help Text*: When true, bot accounts can be used for integrations in Integrations > Bot Accounts. Bot accounts are similar to user accounts except they cannot be used to log in. See documentation to learn more.

NOTE: “documentation” above links to https://mattermost.com/pl/default-bot-accounts

If you have any questions or design clarifications of the proposed improvement, reach out to @jasonblais on GitHub, or jason.blais on https://community.mattermost.com

### PR DESCRIPTION
**Pull request by andresoro**
#### Summary
This PR adds a config flag to enable/disable the creation of bots. With this change, you can enable/disable bots through System Console or REST API. The setting for the System Console will look something like this: 
Screen Shot 2019-05-06 at 1 55 03 PM

#### Ticket Link
#10776

### CODE REVIEW

File: api4/apitestlib.go [hunk 1]
```diff
@@ -212,6 +212,8 @@ func (me *TestHelper) InitBasic() *TestHelper {
 	me.LoginBasic()
 	me.Group = me.CreateGroup()
 
+	me.App.Config().ServiceSettings.CreateBotAccounts = model.NewBool(true)
```
Comment (lieut-data) [inline api4/apitestlib.go]: I don't think this is something we should configure at this level -- if tests need it set one way or another, let's make it explicit in each affected test.

File: i18n/en.json [hunk 2]
```diff
@@ -131,6 +131,10 @@
     "id": "api.admin.upload_brand_image.too_large.app_error",
     "translation": "Unable to upload file. File is too large."
   },
+  {
+    "id": "api.bot.create_disabled",
+    "translation": "Creating of bots disabled in config"
```
Comment (lieut-data) [inline i18n/en.json]: Suggest:
```suggestion
    "translation": "Bot creation has been disabled."
```

File: model/config.go [hunk 3]
```diff
@@ -293,6 +293,7 @@ type ServiceSettings struct {
 	EnableEmailInvitations                            *bool
 	ExperimentalLdapGroupSync                         *bool
 	DisableBotsWhenOwnerIsDeactivated                 *bool `restricted:"true"`
+	CreateBotAccounts                                 *bool
```
Comment (lieut-data) [inline model/config.go]: @crspeller, should this be something that's restricted for Mattermost Cloud?

**Comment by lieut-data** [comment]
@andresoro, could you also investigate adding unit tests to verify this behaviour? Something in `api4/bot_test.go` should suffice.

File: api4/apitestlib.go (hunk 1 above)
Comment (andresoro) [reply api4/apitestlib.go]: Yes this makes sense. Also, undoing this change will allow for proper unit tests.

File: model/config.go (hunk 3 above)
Comment (crspeller) [reply model/config.go]: Ideally no. Currently only system admins can create bots even when this setting is enabled. When it gets added to the permissions UI it should be able to be controlled from there.

**Comment by andresoro** [comment]
Went ahead and added the new unit test under the CreateBot test in bot_test.go and enabled the creation of bot flag in each test that needed it.

**Comment by andresoro** [comment]
Thanks for the reply. I remember seeing that UpdateConfig method when I was modifying model/config.go and it definitely makes more sense to use that. I updated the test cases accordingly.

**Comment by lieut-data** [comment]
@crspeller, I'll leave this to you to merge, since I know there's some outstanding discussion on https://github.com/mattermost/mattermost-webapp/pull/2751 as to what we're actually constraining here.

**Comment by jasonblais** [comment]
@lieut-data As discussed, we'll constrain bot account creation, so this PR should be ready to merge.

**Comment by jasonblais** [comment]
Hey @andresoro - thanks again for the contribution! There was a proposal to rename the config setting from `CreateBotAccounts` to `EnableBotAccountCreation` and updating this name throughout the files (server and webapp). This would help make it more consistent with the naming convention we've used for other similar settings.

Wondering if you were open to help with that change? Sorry that we missed on the original review!

(I created a ticket for it for reference as well: https://mattermost.atlassian.net/browse/MM-15662)

**Comment by andresoro** [comment]
Hey @jasonblais I opened that change up at #10875.

**Comment by jasonblais** [comment]
Thank you!!

**Pull request merged**